### PR TITLE
[MRG+2]Add new deduplication algorithm

### PR DESCRIPTION
**Pull request by floriangardin**
Removing old similarity threshold based deduplication and replacing it with a new algorithm. The algorithm construct a tree that aim to split rules into different homogeneous subsets. Two rules are considered homogeneous if they are semantically near.

### CODE REVIEW

**Comment by datajms** [comment]
LGTM
I'll try it out and make a notebook to explain how to use the max_depth_duplication.

File: skrules/skope_rules.py [hunk 1]
```diff
@@ -61,10 +54,17 @@ class SkopeRules(BaseEstimator):
...
-    max_depth : integer or None, optional (default=3)
+    max_depth : integer or List or None, optional (default=3)
         The maximum depth of the decision trees. If None, then nodes are
         expanded until all leaves are pure or until all leaves contain less
         than min_samples_split samples.
+        If an iterable is passed, you will train n_estimators
+        for each tree depth. It allows you to create and compare
+        rules of different length.
+
+    max_depth_duplication : integer or None, optional (default=3)
```
Comment (ngoix) [inline skrules/skope_rules.py]: default does not match `__init__`, and if None is a possible value it should be a default. So this line has to be
`max_depth_duplication : integer, optional`

File: skrules/skope_rules.py [hunk 2]
```diff
@@ -613,3 +590,70 @@ def _eval_rule_perf(self, rule, X, y):
             return (0, 0)
         pos = y[y > 0].sum()
         return y_detected.mean(), float(true_pos) / pos
+
+    def deduplicate(self, rules):
+        return [max(rules_set, key=self.f1_score) for rules_set in self._find_similar_rulesets(rules)]
```
Comment (ngoix) [inline skrules/skope_rules.py]: this line is too long, please run pep8

File: skrules/skope_rules.py [hunk 3]
```diff
@@ -357,34 +358,10 @@ def fit(self, X, y, sample_weight=None):
...
-                if float(size_intersection)/size_union > self.similarity_thres:
-                    omit_these_rules_list.append(j)
-
-        self.rules_ = [self.rules_[i] for i in range(
-            len(self.rules_)) if i not in omit_these_rules_list]
-
+        # count representation of feature
+        if self.max_depth_duplication is not None:
+            self.rules_ = self.deduplicate(self.rules_)
+            # TODO : Factorize disjoints performing rules (ex : c0 > 0 and c1 > 1  , c0 > 0 and c1 <= 1)
```
Comment (ngoix) [inline skrules/skope_rules.py]: line too long

File: skrules/skope_rules.py [hunk 4]
```diff
@@ -613,3 +590,70 @@ def _eval_rule_perf(self, rule, X, y):
             return (0, 0)
         pos = y[y > 0].sum()
         return y_detected.mean(), float(true_pos) / pos
+
+    def deduplicate(self, rules):
+        return [max(rules_set, key=self.f1_score) for rules_set in self._find_similar_rulesets(rules)]
+
+    def _find_similar_rulesets(self, rules):
+        """Create clusters of rules using a decision tree based on the terms of the rules
```
Comment (ngoix) [inline skrules/skope_rules.py]: too long

File: skrules/skope_rules.py [hunk 5]
```diff
@@ -613,3 +590,70 @@ def _eval_rule_perf(self, rule, X, y):
...
+                    rules_splitted[0].append(rule)
+                elif (most_represented_term + ' >') in rule[0]:
+                    rules_splitted[1].append(rule)
+                else:
+                    rules_splitted[2].append(rule)
+
+            # Choose best term
+            return [split_with_best_feature(ruleset,
+                                            depth-1,
+                                            exceptions=exceptions+[most_represented_term])
```
Comment (ngoix) [inline skrules/skope_rules.py]: ...

File: skrules/skope_rules.py [hunk 6]
```diff
@@ -613,3 +590,70 @@ def _eval_rule_perf(self, rule, X, y):
...
+                for rules_child in rules:
+                    breadth_first_search(rules_child, leaves=leaves)
+            return leaves
+        leaves = []
+        res = split_with_best_feature(rules, self.max_depth_duplication)
+        breadth_first_search(res, leaves=leaves)
+        return leaves
+
+    def f1_score(self, x):
+        return 2 * x[1][0] * x[1][1] / (x[1][0] + x[1][1]) if (x[1][0] + x[1][1]) > 0 else 0
```
Comment (ngoix) [inline skrules/skope_rules.py]: add a new line at the end of file

File: skrules/tests/test_skope_rules.py [hunk 7]
```diff
@@ -76,12 +79,6 @@ def test_skope_rules_error():
     y = iris.target
     y = (y != 0)
 
-    # Test similarity_thres:
```
Comment (ngoix) [inline skrules/tests/test_skope_rules.py]: add similar test on the new parameter for invalid values?

File: skrules/tests/test_skope_rules.py [hunk 8]
```diff
@@ -170,3 +183,45 @@ def test_performances():
...
+    assert_equal(idx_bags_rules[0], idx_bags_rules[1])
+    assert_not_equal(idx_bags_rules[0], idx_bags_rules[2])
+    # Assert the best rules are kept
+    final_rules = sk.deduplicate(rules)
+    assert_in(rules[0], final_rules)
+    assert_in(rules[2], final_rules)
+    assert_not_in(rules[3], final_rules)
+
+
+def test_f1_score():
```
Comment (ngoix) [inline skrules/tests/test_skope_rules.py]: nice

**Comment by ngoix** [comment]
nice PR, lgtm apart from my comments

**Comment by datajms** [comment]
LGTM. I'm merging. Good job Florian!

**Pull request merged**